opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt

```diff
@@ -0,0 +1,571 @@
+        val data = media.split('\t')
+        if (data.size != 2) {
+            OpacApi.ProlongResult(OpacApi.MultiStepResult.Status.ERROR, "internal error")
+        }
+        return try {
+            if (data[0] == "FL") {
+                val formBody = FormBody.Builder()
+                        .add("bc", data[1])
+                        .add("uid", account.name)
+                        .add("clang", "DE")
+                        .add("action", "send")
+                        .build()
+                val result = httpPost(illRenewUrl, formBody, ENCODING).html
+                OpacApi.ProlongResult(OpacApi.MultiStepResult.Status.OK, result.select("p:last-of-type").text())
+            } else {
+                requestAccount(account, "renew", mapOf("tx_slubaccount_account[renewals][0]" to data[1]))
+                OpacApi.ProlongResult(OpacApi.MultiStepResult.Status.OK)
+            }
+        } catch (e: Exception) {
+            OpacApi.ProlongResult(OpacApi.MultiStepResult.Status.ERROR, e.message)
+        }
+    }
+
+    override fun prolongAll(account: Account, useraction: Int, selection: String?): OpacApi.ProlongAllResult {
+        return OpacApi.ProlongAllResult(OpacApi.MultiStepResult.Status.UNSUPPORTED)
+    }
+
+    override fun cancel(media: String, account: Account, useraction: Int, selection: String?): OpacApi.CancelResult {
+        return try {
+            requestAccount(account, "delete", mapOf("tx_slubaccount_account[delete][0]" to media))
+            OpacApi.CancelResult(OpacApi.MultiStepResult.Status.OK)
+        } catch (e: Exception) {
+            OpacApi.CancelResult(OpacApi.MultiStepResult.Status.ERROR, e.message)
+        }
+    }
+
+    override fun account(account: Account): AccountData {
+        val json = requestAccount(account, "account")
+        return parseAccountData(account, json)
+    }
+
+    internal fun parseAccountData(account: Account, json: JSONObject): AccountData {
+        val fmt = DateTimeFormat.shortDate()
+        fun getReservations(items: JSONObject?): MutableList<ReservedItem> {
+            val types = listOf("hold", "request_ready", "readingroom", "request_progress", "reserve")
+            // "requests" is a copy of "request_ready" + "readingroom" + "request_progress"
+            val reservationsList = mutableListOf<ReservedItem>()
+            for (type in types) {
+                items?.optJSONArray(type)?.let {
+                    for (i in 0 until it.length()) {
+                        reservationsList.add(it.getJSONObject(i).let {
+                            ReservedItem().apply {
+                                title = it.optString("about")
+                                author = it.optJSONArray("X_author")?.optString(0)
+                                //id = it.optString("label")  // TODO: get details from here via /bc --> redirects to /id, from there get the proper id
```

Comment:
This is an excellent idea 👍!  I was already a bit unhappy with having to resolve the barcode (bc) to id for every lent or reserved item in every account request just for the relatively unlikely case that it's really going to be used, i.e. that the user wants to look up the catalog entry for an item he has already lent or reserved.  
So I'd need to add a prefix to the id field (`id`, `bc` or maybe `rsn`) and then, if this prefix is not `id`, get the `id` from `bc` in `getResultById` by intercepting the redirection. In `getShareUrl` (I guess you meant this by `getShareLink`) I could simply return a `bc` or `rsn` link instead of an `id` link.  
I'm not sure how fast I can do this, so maybe we can leave this for later, but I'm going to implement it this way.

Reply:
Yes, that's what I meant :) Sure, we can leave this for later.